util/src/zip.rs

```diff
@@ -12,133 +12,76 @@
+use std::io::{self, BufReader, BufWriter, Write};
+use std::path::{Path, PathBuf};
+use std::thread;
 
-use self::zip_rs::result::{ZipError, ZipResult};
 use self::zip_rs::write::FileOptions;
 use zip as zip_rs;
 
-/// Compress a source directory recursively into a zip file.
-/// Permissions are set to 644 by default to avoid any
-/// unwanted execution bits.
-pub fn compress(src_dir: &Path, dst_file: &File) -> ZipResult<()> {
-	if !Path::new(src_dir).is_dir() {
-		return Err(ZipError::Io(io::Error::new(
-			io::ErrorKind::Other,
-			"Source must be a directory.",
-		)));
-	}
+/// Create a zip archive from source dir and list of relative file paths.
+/// Permissions are set to 644 by default.
+pub fn create_zip(dst_file: &File, src_dir: &Path, files: Vec<PathBuf>) -> io::Result<()> {
+	let mut writer = {
+		let zip = zip_rs::ZipWriter::new(dst_file);
+		BufWriter::new(zip)
+	};
 
 	let options = FileOptions::default()
 		.compression_method(zip_rs::CompressionMethod::Stored)
 		.unix_permissions(0o644);
 
-	let mut zip = zip_rs::ZipWriter::new(dst_file);
-	let walkdir = WalkDir::new(src_dir.to_str().unwrap());
-	let it = walkdir.into_iter();
-
-	for dent in it.filter_map(|e| e.ok()) {
-		let path = dent.path();
-		let name = path
-			.strip_prefix(Path::new(src_dir))
-			.unwrap()
-			.to_str()
-			.unwrap();
-
-		if path.is_file() {
-			zip.start_file(name, options)?;
-			let mut f = File::open(path)?;
-			io::copy(&mut f, &mut zip)?;
+	for x in &files {
+		let file_path = src_dir.join(x);
+		if let Ok(file) = File::open(file_path.clone()) {
+			info!("compress: {:?} -> {:?}", file_path, x);
+			writer.get_mut().start_file_from_path(x, options)?;
+			io::copy(&mut BufReader::new(file), &mut writer)?;
+			// Flush the BufWriter after each file so we start then next one correctly.
+			writer.flush()?;
 		}
 	}
 
-	zip.finish()?;
+	writer.get_mut().finish()?;
 	dst_file.sync_all()?;
 	Ok(())
 }
 
-/// Decompress a source file into the provided destination path.
-pub fn decompress<R, F>(src_file: R, dest: &Path, expected: F) -> ZipResult<usize>
-where
-	R: io::Read + io::Seek + panic::UnwindSafe,
-	F: Fn(&Path) -> bool + panic::UnwindSafe,
-{
-	let mut decompressed = 0;
-
-	// catch the panic to avoid the thread quit
-	panic::set_hook(Box::new(|panic_info| {
-		error!(
-			"panic occurred: {:?}",
-			panic_info.payload().downcast_ref::<&str>().unwrap()
-		);
-	}));
-	let result = panic::catch_unwind(move || {
-		let mut archive = zip_rs::ZipArchive::new(src_file)?;
+/// Extract a set of files from the provided zip archive.
+pub fn extract_files(from_archive: File, dest: &Path, files: Vec<PathBuf>) -> io::Result<()> {
+	let dest: PathBuf = PathBuf::from(dest);
+	let files: Vec<_> = files.iter().cloned().collect();
+	let res = thread::spawn(move || {
+		let mut archive = zip_rs::ZipArchive::new(from_archive).expect("archive file exists");
+		for x in files {
+			if let Ok(file) = archive.by_name(x.to_str().expect("valid path")) {
```

Comment:
Nitpick: `valid path` and all below may look confusing in the log file, looks good in the source code though